streams/dshr_strdata_mod.F90

```diff
@@ -1225,123 +1422,230 @@ subroutine shr_strdata_readstrm(sdat, ns, stream, stream_mesh, &
+                      else
+                         dataptr2d(l,n) = r8fill
+                      endif
+                   enddo
+                end do
+             else
+                do l = 1,stream_nlev
+                   do n = 1,size(dataptr2d, dim=2)
+                      dataptr2d(l,n) = real(data_real2d(n,l), kind=r8)
+                   end do
+                end do
+             end if
+          else ! stream_nlev == 1
+             if (per_stream%stream_pio_iodesc_set) then
+                call pio_read_darray(pioid, varid, per_stream%stream_pio_iodesc, data_real1d, rcode)
+             else
+                rcode = pio_get_var(pioid, varid,start=(/1,1,nt/),count=(/1,1,1/), ival=data_real1d)
+             endif
+             if ( rcode /= PIO_NOERR ) then
+                call shr_sys_abort(' ERROR: reading in variable: '// trim(per_stream%fldlist_stream(nf)))
+             end if
+             if (handlefill) then
+                do n=1,size(dataptr1d)
+                   if(.not. shr_infnan_isnan(data_real1d(n)) .and. data_real1d(n) .ne. fillvalue_r4) then
+                      dataptr1d(n) = real(data_real1d(n), kind=r8)
+                   else
+                      dataptr1d(n) = r8fill
+                   endif
+                enddo
+             else
+                dataptr1d(:) = real(data_real1d(:),kind=r8)
+             endif
           end if
-          if(handlefill) then
-             do n=1,size(dataptr)
-                if(.not. shr_infnan_isnan(data_real(n)) .and. data_real(n) .ne. fillvalue_r4) then
-                   dataptr(n) = real(data_real(n), kind=r8)
-                else
-                   dataptr(n) = r8fill
-                endif
-             enddo
-          else
-             dataptr(:) = real(data_real(:),kind=r8)
-          endif
+
        else if (pio_iovartype == PIO_DOUBLE) then
-          if (pio_iodesc_set) then
-             call pio_read_darray(pioid, varid, pio_iodesc, dataptr, rcode)
-          else
-             rcode = pio_get_var(pioid, varid,start=(/1,1,nt/),count=(/1,1,1/), ival=dataptr)
-          endif
-          if ( rcode /= PIO_NOERR ) then
-             call shr_sys_abort(' ERROR: reading in variable: '// trim(fldlist_stream(nf)))
+          ! -----------------------------
+          ! pio_iovartype is PIO_DOUBLE
+          ! -----------------------------
+          if (stream_nlev > 1) then
+             if (per_stream%stream_pio_iodesc_set) then
+                call pio_read_darray(pioid, varid, per_stream%stream_pio_iodesc, data_dbl2d, rcode)
+             else
+                rcode = pio_get_var(pioid, varid,start=(/1,1,1,nt/), count=(/1,1,1,1/), ival=data_dbl2d)
+             end if
+             if ( rcode /= PIO_NOERR ) then
+                call shr_sys_abort(' ERROR: reading in 2d double variable: '// trim(per_stream%fldlist_stream(nf)))
+             end if
+             if (handlefill) then
+                do l = 1,stream_nlev
```

Comment:
Very hard to distinguish the l in this loop to the 1 in line 1505.

Reply:
same comment as above.